core/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java

```diff
@@ -189,22 +208,34 @@ public static Builder builder(Client client, Listener listener) {
 
     private volatile boolean closed = false;
 
-    BulkProcessor(Client client, BackoffPolicy backoffPolicy, Listener listener, @Nullable String name, int concurrentRequests, int bulkActions, ByteSizeValue bulkSize, @Nullable TimeValue flushInterval) {
+    BulkProcessor(BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer, BackoffPolicy backoffPolicy, Listener listener,
+                  @Nullable String name, int concurrentRequests, int bulkActions, ByteSizeValue bulkSize, @Nullable TimeValue flushInterval,
+                  @Nullable ThreadPool threadPool, Settings settings) {
         this.bulkActions = bulkActions;
         this.bulkSize = bulkSize.getBytes();
-
         this.bulkRequest = new BulkRequest();
-        this.bulkRequestHandler = (concurrentRequests == 0) ? BulkRequestHandler.syncHandler(client, backoffPolicy, listener) : BulkRequestHandler.asyncHandler(client, backoffPolicy, listener, concurrentRequests);
 
-        if (flushInterval != null) {
-            this.scheduler = (ScheduledThreadPoolExecutor) Executors.newScheduledThreadPool(1, EsExecutors.daemonThreadFactory(client.settings(), (name != null ? "[" + name + "]" : "") + "bulk_processor"));
-            this.scheduler.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
-            this.scheduler.setContinueExistingPeriodicTasksAfterShutdownPolicy(false);
-            this.scheduledFuture = this.scheduler.scheduleWithFixedDelay(new Flush(), flushInterval.millis(), flushInterval.millis(), TimeUnit.MILLISECONDS);
+        BiFunction<TimeValue, Runnable, ScheduledFuture<?>> scheduleFn;
+        if (threadPool == null) {
+            ThreadFactory threadFactory = EsExecutors.daemonThreadFactory(settings, (name != null ? "[" + name + "]" : "") + "bulk_processor");
+            ScheduledThreadPoolExecutor scheduledExecutor = (ScheduledThreadPoolExecutor) Executors.newScheduledThreadPool(1, threadFactory);
+            scheduledExecutor.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+            scheduledExecutor.setContinueExistingPeriodicTasksAfterShutdownPolicy(false);
+            this.schedulerToStop = scheduledExecutor;
+            scheduleFn = Retry.fromScheduledExecutorService(scheduledExecutor);
```

Comment:
I tend to agree with Simon on requiring a ThreadPool. We can have a helper in the high level client to create one. Otherwise the code here gets too complex and I am not sure it is worth it.